infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+                        // If timed out, return a exceptionally completed with the timeout.
+                        // Because in Log Replication, messages are sent to the leader node, the connection future
+                        // represents a connection to the leader.
+                        try {
+                            sessionToLeaderConnectionFuture.get(session).get(timeoutResponse, TimeUnit.MILLISECONDS);
+                        } catch (InterruptedException e) {
+                            throw new UnrecoverableCorfuInterruptedError(e);
+                        } catch (TimeoutException | ExecutionException te) {
+                            cf.completeExceptionally(te);
+                            return cf;
+                        }
+                    }
+
+                    if(sessionManager.getOutgoingSessions().contains(session)) {
+                        CorfuLogReplicationRuntime runtimeFSM = sessionToRuntimeFSM.get(session);
+                        // Get Remote Leader
+                        if (runtimeFSM.getRemoteLeaderNodeId().isPresent()) {
+                            nodeId = runtimeFSM.getRemoteLeaderNodeId().get();
+                        } else {
+                            log.error("Leader not found to remote cluster {}", sessionToRemoteClusterDescriptor.get(session).getClusterId());
+                            runtimeFSM.input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS));
+                            throw new ChannelAdapterException(
+                                    String.format("Leader not found to remote cluster %s", sessionToRemoteClusterDescriptor.get(session).getClusterId()));
+                        }
+                    }
+                }
+
+                // In the case the message is intended for a specific endpoint, we do not
+                // block on connection future, this is the case of leader verification.
+                if(isConnectionInitiator(session)) {
+                    clientChannelAdapter.send(nodeId, getRequestMsg(header.build(), payload));
+                } else {
+                    serverChannelAdapter.send(getRequestMsg(header.build(), payload));
+                }
+            } catch (NetworkException ne) {
+                log.error("Caught Network Exception while trying to send message to remote leader {}", nodeId);
+                if(sessionManager.getOutgoingSessions().contains(session)) {
+                    sessionToRuntimeFSM.get(session).input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.ON_CONNECTION_DOWN,
+                            nodeId));
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.ON_CONNECTION_DOWN,
```

Comment:
if a node is a Sink(from replication perspective) and a connection initiator, the runtime FSM will not be running on it, right?  How will this event be processed?

Reply:
VerifyRemoteSourceLeadership was using the runtimeEvent, and was not putting the event in the runtimeFSM queue.
After Chris comment (https://github.com/CorfuDB/CorfuDB/pull/3560#discussion_r1149877895), I have created another enum for Sink.
It will be present in the next push.